apps/nextjs/components/TopBar.tsx

```diff
@@ -0,0 +1,57 @@
+import {
+  CommunitiesLink,
+  DiscordLogo,
+  GitHubLogo,
+  PackageUploadLink,
+  TopBar as TopBarComponent,
+  TopBarMenu,
+  TopBarMenuButton,
+} from "@thunderstore/components";
+import Router from "next/router";
+
+import { generateRandomString } from "utils/string";
+
+export const TopBar: React.FC = () => {
+  return (
+    <TopBarComponent>
+      <PackageUploadLink variant="ts.topBar">Upload</PackageUploadLink>
+      <CommunitiesLink variant="ts.topBar">Browse</CommunitiesLink>
+      <TopBarMenu label="Login with…">
+        <TopBarMenuButton
+          icon={DiscordLogo}
+          label="Discord"
+          onClick={onDiscordClick}
+        />
+        <TopBarMenuButton
+          icon={GitHubLogo}
+          label="GitHub"
+          onClick={onGitHubClick}
+        />
+      </TopBarMenu>
+    </TopBarComponent>
+  );
+};
+
+const onDiscordClick = () => {
+  const [state, redirectUrl] = getStateAndRedirectUrl("discord");
+  const clientId = process.env.NEXT_PUBLIC_DISCORD_ID;
+  Router.push(
+    `https://discord.com/api/oauth2/authorize?client_id=${clientId}&redirect_uri=${redirectUrl}&state=${state}&response_type=code&scope=email%20identify`
+  );
+};
+
+const onGitHubClick = () => {
+  const [state, redirectUrl] = getStateAndRedirectUrl("github");
+  const clientId = process.env.NEXT_PUBLIC_GITHUB_ID;
+  Router.push(
+    `https://github.com/login/oauth/authorize?client_id=${clientId}&redirect_uri=${redirectUrl}&state=${state}&response_type=code&scope=user:email,read:user`
+  );
+};
+
+/** Return CSRF protection token and post-login redirect URL */
+const getStateAndRedirectUrl = (provider: string) => {
+  const state = generateRandomString();
+  window.localStorage.setItem(`${provider}LoginState`, state);
```

Comment:
Be helpful how?

Reply:
Just as an abstraction to separate the logic of what we want to do away from the actual storage implementation. I could see the method of storage being changed at some point, even if it'd just be the way we operate the same storage (i.e. key being renamed or us storing a json blob instead of a raw value).

Same reason we have an ORM layer in Django instead of writing raw DB queries

Reply:
Added [utils/storage](https://github.com/thunderstore-io/thunderstore-ui/pull/404/commits/0498117426766569ce94a51cd56871c969aeac43#diff-3a72004aac89f3edfa66b9f08f16edd438adee1c674dd0039ea65ca139959f8f)